llvm/include/llvm/Support/TrailingObjects.h

```diff
@@ -301,6 +302,41 @@ class TrailingObjects : private trailing_objects_internal::TrailingObjectsImpl<
         static_cast<BaseTy *>(this), TrailingObjectsBase::OverloadToken<T>());
   }
 
+  // getTrailingObjects() specialization for a single trailing type.
+  using FirstTrailingType =
+      typename std::tuple_element_t<0, std::tuple<TrailingTys...>>;
+
+  const FirstTrailingType *getTrailingObjects() const {
+    static_assert(sizeof...(TrailingTys) == 1,
+                  "Can use non-templated getTrailingObjects() only when there "
+                  "is a single trailing type");
+    return getTrailingObjects<FirstTrailingType>();
+  }
+
+  FirstTrailingType *getTrailingObjects() {
+    static_assert(sizeof...(TrailingTys) == 1,
+                  "Can use non-templated getTrailingObjects() only when there "
+                  "is a single trailing type");
+    return getTrailingObjects<FirstTrailingType>();
+  }
+
+  // Functions that return the trailing objects as ArrayRefs.
+  template <typename T> MutableArrayRef<T> getTrailingObjects(size_t N) {
+    return {getTrailingObjects<T>(), N};
```

Comment:
I think you cannot construct an ArrayRef to an array that you construct on the stack? That looks like a use-after-return to me.

Reply:
`getTrailingObjects<T>()` returns a pointer to the heap allocated TrailingObject, so this is just wrapping up that heap pointer in an ArrayRef.

Reply:
See my earlier PR that shows existing code that does this in all places (and which I am planning to replace with this API): https://github.com/llvm/llvm-project/pull/138554

Reply:
Ah, can you please use a more explicit constructor? I was confused for a second and thought we are constructing an array of two elements.

Reply:
Like ArrayRef(getTrailingObjects<T>(), N)? Sure I can do that.

Reply:
yeah, i think that's more explicit and avoids this confusion. i left the comment when i just quickly reskimmed the CL and `{...}` is mostly hardwired to element initialization for me.

Reply:
Done.